server/dirstore.go

```diff
@@ -0,0 +1,706 @@
+		if strings.HasSuffix(path, extension) && filepath.Dir(path) == store.directory {
+			store.addFile(path)
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	quitError := errors.New("abort")
+	errCb := store.errorOccurred
+	chCb := store.changed
+	wg := store.done
+	quit := store.expiration.quit
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			select {
+			case <-time.After(time.Second):
+			case <-quit:
+				return
+			}
+			// check for additions or modifications
+			err := filepath.Walk(store.directory, func(path string, info os.FileInfo, err error) (retErr error) {
+				if err != nil {
+					return err
+				}
+				if strings.HasSuffix(path, extension) && filepath.Dir(path) == store.directory {
+					if changed, err := store.addFile(path); err != nil {
+						errCb(err)
+					} else if changed {
+						pubKey := strings.Replace(filepath.Base(path), ".jwt", "", -1)
+						chCb(pubKey)
+					}
+				}
+				select {
+				case <-time.After(time.Second):
+				case <-quit:
+					return quitError
+				}
+				return
+			})
+			if err != nil && err != quitError {
+				errCb(err)
+			}
+		}
+	}()
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			select {
+			case <-time.After(time.Second):
+			case <-quit:
+				return
+			}
+			store.Lock() // access of e is save, pop will nil out next
+			for e := store.expiration.lru.Front(); e != nil; e = e.Next() {
+				pubKey := e.Value.(*JWTItem).publicKey
+				path := store.pathForKey(pubKey)
+				if _, err := os.Stat(path); os.IsNotExist(err) {
+					store.expiration.UnTrack(pubKey)
+				}
+				store.Unlock()
+				select {
+				case <-time.After(time.Second):
+				case <-quit:
+					return
+				}
+				store.Lock()
+			}
+			store.Unlock()
+		}
+	}()
+	return nil
+}
+
+// Load checks the memory store and returns the matching JWT or an error
+// Assumes lock is NOT held
+func (store *DirJWTStore) load(publicKey string) (string, error) {
+	store.Lock()
+	defer store.Unlock()
+	if path := store.pathForKey(publicKey); path == "" {
+		return "", fmt.Errorf("invalid public key")
+	} else if data, err := ioutil.ReadFile(path); err != nil {
+		return "", err
+	} else {
+		if store.expiration != nil {
+			store.expiration.UpdateTrack(publicKey)
+		}
+		return string(data), nil
+	}
+}
+
+// Save puts the JWT in a map by public key and performs update callbacks
+// Assumes lock is NOT held
+func (store *DirJWTStore) save(publicKey string, theJWT string) error {
+	store.Lock()
+	if store.readonly {
+		store.Unlock()
+		return fmt.Errorf("store is read-only")
+	}
+	path := store.pathForKey(publicKey)
+	if path == "" {
+		store.Unlock()
+		return fmt.Errorf("invalid public key")
+	}
+	dirPath := filepath.Dir(path)
+	if _, err := validateDirPath(dirPath); err != nil {
+		if err := os.MkdirAll(dirPath, 0755); err != nil {
+			store.Unlock()
+			return err
+		}
+	}
+	changed, err := store.write(path, publicKey, theJWT)
+	cb := store.changed
+	store.Unlock()
+	if changed && cb != nil {
```

Comment:
Should we do cb if we had an err?

Reply:
changed will never be true when an error is returned. 
Added that to the comments of write